### PR TITLE
Mask MIMO MTP prediction paths that cross modality tokens

### PR DESCRIPTION
**Pull request by kevjshih**
- [x] I, the PR author, have personally reviewed every line of this PR.

# What does this PR do?

Prevents multimodal placeholder tokens from becoming accidental learning signals for multi-token prediction (MTP).

This is a correctness fix for VLM pretraining with MTP. It does not enable or supervise vision-token prediction: VLM training intentionally masks vision-token targets, and generation never asks MTP to produce image tokens.

The change covers every special token ID configured by MIMO, including image and audio placeholders.

## Why is this needed?

MTP predicts several future text tokens by repeatedly adding the next teacher-forced token's embedding to the prediction path. Existing target masking prevents direct supervision to predict a modality placeholder, but that alone does not stop the placeholder from being used as an input by a later MTP depth.

For example:

```text
"look" -> "at" -> <image> -> "it" -> "shines"
```

For the position at `"look"`, the prediction path is:

1. The normal next-token head predicts `"at"`.
2. MTP depth 1 consumes the known token `"at"` and is aligned with `<image>` as its target. Existing target masking correctly gives this position zero loss; MTP is not trained to predict the image token.
3. MTP depth 2 then consumes the known `<image>` placeholder and is aligned with the text target `"it"`. The target is text, but this prediction is still invalid because a vision placeholder was used as an ordinary token-embedding input.
4. MTP depth 3 consumes `"it"` and is aligned with `"shines"`. Both are text, but this causal path already crossed `<image>` at depth 2, so it must remain masked.

This is the multi-depth corner case: even when a deeper MTP head currently consumes and predicts text, its path may have consumed a vision token at an earlier depth. Validity therefore has to accumulate across MTP depths rather than being recomputed from only the current target or input.

Training a text prediction from an MTP state that consumed a vision placeholder ID as an ordinary token embedding is invalid supervision. Input and output embeddings are also commonly tied, so an input-side gradient can modify the shared placeholder row. Neither behavior is intended; this change does not make image-token generation part of the MTP objective.

This PR excludes each affected per-position prediction path after it consumes a modality placeholder and prevents invalid MTP decoder embeddings from receiving input-side gradients. The MTP module still executes over the tensor for efficiency; the affected positions contribute neither loss nor acceptance metrics.

## What changed?

- Reuse MIMO's text-validity mask as the MTP conditioning mask, and build it only on ranks that actually execute MTP.
- Preserve token IDs on pipeline stages that own MTP so shifted-token embeddings can be constructed when hidden states arrive from an earlier stage.
- Apply the same context-parallel partition to token IDs and position IDs as the corresponding hidden states.
- Thread the conditioning mask through GPT, hybrid, and fine-grained MTP execution paths.
- Roll token IDs and their validity together, avoiding an additional context-parallel boundary exchange.
- Detach invalid shifted-token embeddings while preserving their forward values, so later causal losses cannot update tied placeholder rows.
- Roll and accumulate the conditioning mask for each MTP step, and use the combined mask for loss normalization and acceptance metrics.
- Preserve the existing no-mask token-count fast path for callers that do not provide an MTP conditioning mask.
- Pass MTP's process groups and pipeline rank from the model, avoiding global process-state lookups in the normal execution path.

## Issue tracking

Related to #6010.

## Testing

Pre-fix regression checks reproduced each reviewed failure:

- PP=2: the last stage reached MTP with `input_ids=None` and failed its embedding lookup.
- CP=2: MTP concatenation failed because global token IDs produced sequence lengths of 16/8 while CP-local hidden states had lengths of 8/4 for TP=1/2.
- Hybrid: enabled MTP execution received no conditioning mask; disabled controls passed.
- Gradient: all 32 elements of the invalid shared embedding row received nonzero input-side gradient.

Post-fix validation:

- Black, isort, Pylint (10.00/10), and Ruff: passed.
- Mypy comparison against the same `main` baseline: no new diagnostics.
- Final distributed TP/CP/PP, gradient, packed-sequence, and MTP placement suite after merging the latest `main` (4 ranks): 93 passed and 15 expected eight-rank-only cases skipped per rank.
- Expanded coverage includes image and audio placeholders, 2D and 3D position IDs, packed CP boundaries, MTP depths 1 and 2, and a check that the conditioning mask adds no separate CP boundary exchange.
- `git diff --check`: passed.

### Pre-checks

- [x] I have added relevant unit tests
- [ ] I have added relevant functional tests (not applicable; this is unit-level masking and parallelism behavior)
- [x] I have added proper typing to my code [Typing guidelines](https://github.com/NVIDIA/Megatron-LM/blob/main/docs/contribute/style_guide.md#type-annotations)
- [ ] I have added relevant documentation (not applicable; no user-facing API or configuration change)
- [x] I have run the [autoformatter.sh](https://github.com/NVIDIA/Megatron-LM/blob/main/tools/autoformat.sh) on my PR

### CODE REVIEW

**Comment by kevjshih** [comment]
/ok to test [ee02e97](https://github.com/NVIDIA/Megatron-LM/pull/6770/commits/ee02e978643aa76171f76c751142e7739df58305)

**Comment by kevjshih** [comment]
/ok to test [3f045ce](https://github.com/NVIDIA/Megatron-LM/commit/3f045cea0f0e9a4a5c0eec1fd7548de6ffeed782)

**Comment by kevjshih** [comment]
/ok to test [3feb1a4](https://github.com/NVIDIA/Megatron-LM/pull/6770/commits/3feb1a442d70192645f68fcede759f7bc34dea69)

**Comment by kevjshih** [comment]
/ok to test [1e64de1](https://github.com/NVIDIA/Megatron-LM/pull/6770/commits/1e64de1dbb971ef463b4a45e2b5b873e24f579ae)

File: megatron/core/transformer/multi_token_prediction.py [hunk 1]
```diff
@@ -847,12 +860,28 @@ def get_mtp_ranks(pp_ranks: List[int], config: TransformerConfig) -> List[int]:
...
-def get_mtp_layer_offset(config: TransformerConfig, vp_stage: Optional[int] = None) -> int:
+def get_mtp_layer_offset(
+    config: TransformerConfig, vp_stage: Optional[int] = None, pp_rank: Optional[int] = None
+) -> int:
     """Get the offset of the MTP layer."""
     if config.pipeline_model_parallel_size > 1:
         if config.pipeline_model_parallel_layout:
-            offset = config.pipeline_model_parallel_layout.get_layer_offset(
-                layer_type=LayerType.mtp, vp_stage=vp_stage
+            if pp_rank is None:
```
Comment (yashaswikarnati) [inline megatron/core/transformer/multi_token_prediction.py:876]: just a nit; may be we can use a small helper function for if pp is none block and other assert checks in this file and reuse across all functions for brevity

**Review by yashaswikarnati — APPROVED** [review]

**Review by sancha — APPROVED** [review]

**Review by Phlip79 — APPROVED** [review]
Please add tests for `HybridModel`. Currently the only tests being added are for the deprecated `GPTModel`. Approving conditional on these tests being added.

File: megatron/core/models/common/model_chunk_schedule_plan.py [hunk 2]
```diff
@@ -370,6 +371,7 @@ def __init__(
             extra_block_kwargs: Additional keyword arguments for blocks.
             runtime_gather_output: Whether to gather output at runtime.
             loss_mask (torch.Tensor): Used to mask out some portions of the loss
+            mtp_input_mask (torch.Tensor): Mask of valid MTP conditioning tokens.
```
Comment (jenchen13) [inline megatron/core/models/common/model_chunk_schedule_plan.py]: what values are allowed in this mask and which values correspond to valid vs invalid MTP conditioning tokens? Can this info be added to the docstring?

**Comment by deepakn94** [comment]
/ok to test [ede9181](https://github.com/NVIDIA/Megatron-LM/pull/6770/commits/ede9181d09acef9ccc2f07ee59bb9c1f3d80dd0c)

**Review by jenchen13 — APPROVED** [review]

**Review by maanug-nv — APPROVED** [review]

**Comment by zhongbozhu** [comment]
Essentially now you can predict a <image> token instead of just predicting text token, so the proof for this fix to be working is that you will observe MTP hit rate increase? please correct me if I am wrong.

**Comment by kevjshih** [comment]
@zhongbozhu I don't believe we've directly measured the impact of this with regards to token acceptance rates, but the intention for now is to ensure the MTP head is strictly supervised by full text token sequences.

**Comment by sancha** [comment]
@zhongbozhu  this is correctness bugfix for VLM pretraining + MTP. We don't predict vision tokens, this is intentional, we are never generating image tokens. This change also adds a corner case if we are training with MTP > 1 the MTP head s may take a vision token as INPUT, then also we dont want the MTP head to activate.

**Review by jaredcasper — APPROVED** [review]

**Comment by yashaswikarnati** [comment]
/ok to test [e9ae1b8](https://github.com/NVIDIA/Megatron-LM/pull/6770/commits/e9ae1b8d7aeb17f0aa7d711e86394cbaedfc5371)

**Review by zhongbozhu — APPROVED** [review]

**Comment by yashaswikarnati** [comment]
/ok to test [e86eb72](https://github.com/NVIDIA/Megatron-LM/pull/6770/commits/e86eb72d26e2d99779cd7bff8252a2a6696a8697)

**Comment by kevjshih** [comment]
/ok to test [5781b86](https://github.com/NVIDIA/Megatron-LM/pull/6770/commits/5781b860989b3a58a01143534cf90a912ed7dde8)

**Comment by yashaswikarnati** [comment]
/ok to test [b7f27c7](https://github.com/NVIDIA/Megatron-LM/pull/6770/commits/b7f27c7ccb12a7a5182e2d35cf9f82d978b82e08)
